### PR TITLE
New API for location subscriptions

### PR DESCRIPTION
**Pull request by chia2017**
#### What type of PR is this?
enhancement/feature

#### What this PR does / why we need it:
Create a new API for creating subscriptions for location based events

#### Which issue(s) this PR fixes:
This PR is related to issue #18  Subscription for Location Area 

Fixes #

#### Changelog input

```
 release-note

```

#### Additional documentation 

This section can be blank.

```
docs

```

### CODE REVIEW

File: code/API_definitions/location-subscriptions.yaml
Comment (jlurien) [inline code/API_definitions/location-subscriptions.yaml:380]: Here we may consider to convert this to an array `eventTypes` and allow subscribing to both `AREA_ENTERED` and `AREA_LEFT` for same `eventSubscriptionId, which may be a common case`. Otherwise 2 independent subscriptions are required for client interested in both. If a client is only interested in one of them, or prefer to manage both subscriptions independently, it can do it as well.

File: code/API_definitions/location-subscriptions.yaml
Comment (jlurien) [inline code/API_definitions/location-subscriptions.yaml:26]: We have to choose a more specific name for this API, as in future we are likely to have another location subscription API for geotracking. This one is related to location verification or geofencing use case. There is an open issue to choose this name: #69 
Once chosen, we have to align it here and also in title, descriptions, etc

File: code/API_definitions/location-subscriptions.yaml
Comment (jlurien) [inline code/API_definitions/location-subscriptions.yaml:4]: As part of the API spec, we have to provide the API doc, as it is already done for location-verification.yaml. 
This can be completed in a separated PR once we agree on the content of this one.

**Review by jlurien — COMMENTED** [review]
It is very well aligned with Commonalities, so nothing to comment there.
- Area and Circle models are being changed in another ongoing PR (#77). We will have to align there also here. We may discuss at this point if it makes sense to have a separated file with common schemas for Device and Area
- I would allow to subscribe to both events, for same Device and Area, in the same subscription

File: code/API_definitions/location-subscriptions.yaml
Comment (bigludo7) [reply code/API_definitions/location-subscriptions.yaml:380]: I've slight preference to keep it single (and not array) as we have this pattern in a Device Status P/R.
I imagine having One subscription - one event type will simplify the monetization (and for developer it did not require additional work as we use same api for distinct event)

File: code/API_definitions/location-subscriptions.yaml
Comment (bigludo7) [inline code/API_definitions/location-subscriptions.yaml:200]: We should avoid use of oneOf as stated in https://github.com/camaraproject/Commonalities/pull/22
in order to encourage inheritance rather than polymorphism.

File: code/API_definitions/location-subscriptions.yaml
Comment (bigludo7) [inline code/API_definitions/location-subscriptions.yaml:222]: We must use meter - we have changed from Km to m.
ref: https://github.com/camaraproject/DeviceLocation/issues/60

File: code/API_definitions/location-subscriptions.yaml
Comment (bigludo7) [inline code/API_definitions/location-subscriptions.yaml:223]: Same use meter so...2000

File: code/API_definitions/location-subscriptions.yaml
Comment (bigludo7) [inline code/API_definitions/location-subscriptions.yaml:218]: For the 2 other API we changed to  `center` to make ot more explicit

File: code/API_definitions/location-subscriptions.yaml
Comment (bigludo7) [inline code/API_definitions/location-subscriptions.yaml:465]: As stated in the guideline we should have a notification for SUBSCRIPTION_ENDS

File: code/API_definitions/location-subscriptions.yaml
Comment (bigludo7) [inline code/API_definitions/location-subscriptions.yaml:435]: Not sure I got it - We did not provide capability to subscribe in future so the start date will be always in the past.
I have a poor english grammar but shouldn't be : The time when the location-tracking started. Provided in date-time format.

File: code/API_definitions/location-subscriptions.yaml
Comment (bigludo7) [inline code/API_definitions/location-subscriptions.yaml:440]: So it is always a date in future ? We assume there that a subscription terminated is de facto deleted. 
I'm fine with that but just to be sure we're on the same page.

File: code/API_definitions/location-subscriptions.yaml
Comment (bigludo7) [inline code/API_definitions/location-subscriptions.yaml:41]: We did not have  right now a decision for using call back or providing a standard endpoint for Notification.
There is is an issue in commonalities to discuss this https://github.com/camaraproject/Commonalities/issues/8

**Review by bigludo7 — COMMENTED** [review]
Thanks a lot !
Few comments.
Main one is to have a common CAMARA decision on using callback or specific endpoint. Both works but my developer feedback told me that callback did not work well with code generator.
We need also alignment on subscription starts/expires date.

File: code/API_definitions/location-subscriptions.yaml
Comment (jlurien) [reply code/API_definitions/location-subscriptions.yaml:380]: But I see that in Device Status there is an event `ROAMING_STATUS` that covers the other two `ROAMING_ON` and `ROAMING_OFF`, so in that API it would be redundant. But here it may also common that a developer is interested in both.

Comment (bigludo7) [reply code/API_definitions/location-subscriptions.yaml:380]: Yes - `ROAMING_ON` and `ROAMING_OFF` cover specific UC - we had a discussion in this project about very specific event.

You can have some UC with moving IoT abroad that require an update and then app are interested for `ROAMING_OFF` , while have some IoT for example that must not left national territory (and `ROAMING_ON` is subscribed as an alarm).

`ROAMING_STATUS` is the generic one for UC where both event are important.

File: code/API_definitions/location-subscriptions.yaml
Comment (chia2017) [reply code/API_definitions/location-subscriptions.yaml:4]: Yes, I saw API doc for location-verification.yaml. It means in this PR we do not need to change it? Shall we prepare a draft for this text and then discuss about content of it or how should be the process for that agreement?

File: code/API_definitions/location-subscriptions.yaml
Comment (chia2017) [reply code/API_definitions/location-subscriptions.yaml:222]: Thanks a lot for your nice and positive comments. I can do the small changes and whenever the decision about major things like naming, callback, eventtype(array or single) is finilized (I assume maybe some other people write comments) then I will change them as well.
Thanks.

File: code/API_definitions/location-subscriptions.yaml
Comment (akoshunyadi) [reply code/API_definitions/location-subscriptions.yaml:4]: @chia2017 in this PR we should focus on the technical content and when this PR has been merged we can create a new PR with the documentation.

File: code/API_definitions/location-subscriptions.yaml
Comment (jlurien) [inline code/API_definitions/location-subscriptions.yaml:507]: We have to provide the details of the event for this specific event

File: code/API_definitions/location-subscriptions.yaml
Comment (jlurien) [inline code/API_definitions/location-subscriptions.yaml:495]: We have to provide the details of the event for this specific event

**Review by jlurien — CHANGES_REQUESTED** [review]
Need to define the eventDetail for the events to be notified

File: code/API_definitions/location-subscriptions.yaml
Comment (chia2017) [reply code/API_definitions/location-subscriptions.yaml:218]: I changed the location property inside Circle to center

File: code/API_definitions/location-subscriptions.yaml
Comment (chia2017) [reply code/API_definitions/location-subscriptions.yaml:435]: I have also a poor english grammer, but based on what you explained yes it should be : The time when the location-tracking started. I changed it

File: code/API_definitions/location-subscriptions.yaml
Comment (chia2017) [reply code/API_definitions/location-subscriptions.yaml:440]: I changed like the comment above.

**Comment by jlurien** [comment]
@chia2017 you still have to define the details in the specific events being notified, that is the device and area. 

            eventDetail:
              type: object -> here instead of object

File: code/API_definitions/location-subscriptions.yaml
Comment (chia2017) [reply code/API_definitions/location-subscriptions.yaml:528]: Please check the last commit.

File: code/API_definitions/location-subscriptions.yaml
Comment (jlurien) [inline code/API_definitions/location-subscriptions.yaml:467]: There was a discussion in QoD to favour allOf vs anyOf to model the event, due to better compatibility with some generators and coding languages, you can check the PR here: https://github.com/camaraproject/QualityOnDemand/pull/177

This was proposed by Orange, @bigludo7

File: code/API_definitions/location-subscriptions.yaml
Comment (fernandopradocabrillo) [inline code/API_definitions/location-subscriptions.yaml:393]: @bigludo7 Here we are adding SUBSCRIPTION_ENDS event in the operation to create a subscription and I was checking yesterday's PR from Device Status and we didn't add it there: https://github.com/camaraproject/DeviceStatus/blob/main/code/API_definitions/device-status.yaml#L513

should we align this?

Comment (jlurien) [reply code/API_definitions/location-subscriptions.yaml:393]: Really, I think it make no sense to allow a subscription to SUBSCRIPTION_ENDS event, as this event is triggered when a subscription to another event ends

Comment (akoshunyadi) [reply code/API_definitions/location-subscriptions.yaml:393]: I also think that we don't need that event in the create-subscription, but in the event-notification

File: code/API_definitions/location-subscriptions.yaml
Comment (jlurien) [inline code/API_definitions/location-subscriptions.yaml:401]: This is not correct. allOf is not extending the enum, it just requires that the object validated against the schema is compliant with all of the sub-schemas, and it cannot comply with both enum at the same time. You have to define an enum with the 3 allowed values for the NotifiedEventType.

**Review by jlurien — CHANGES_REQUESTED** [review]
A new adhoc enum with the 3 values have to be defined

File: code/API_definitions/location-subscriptions.yaml
Comment (chia2017) [reply code/API_definitions/location-subscriptions.yaml]: Yaml file is updated.

**Review by jlurien — APPROVED** [review]
LGTM

File: code/API_definitions/location-subscriptions.yaml
Comment (jlurien) [reply code/API_definitions/location-subscriptions.yaml:26]: it seems that `geofencing` is the name with more consensus, so you can apply it

**Review by jlurien — COMMENTED** [review]
Rename api to geofencing

**Comment by chia2017** [comment]
> Rename api to geofencing

Hi, I renamed the api two weeks ago before I went to vacation. So I did not recieve any new comment about review, If you think everything is fine you can merge this PR, but first it need two approvers. Anyway I am available if you have any comment.
Best,
Sadeghi

File: code/API_definitions/geofencing.yaml
Comment (jlurien) [inline code/API_definitions/geofencing.yaml:190]:
```suggestion
    oAuth2AuthorizationCode:
```

File: code/API_definitions/geofencing.yaml
Comment (jlurien) [inline code/API_definitions/geofencing.yaml:92]:
```suggestion
        - Geofencing
```

File: code/API_definitions/geofencing.yaml
Comment (jlurien) [inline code/API_definitions/geofencing.yaml:119]:
```suggestion
        - Geofencing
```

File: code/API_definitions/geofencing.yaml
Comment (jlurien) [inline code/API_definitions/geofencing.yaml:151]:
```suggestion
        - Geofencing
```

File: code/API_definitions/geofencing.yaml
Comment (jlurien) [inline code/API_definitions/geofencing.yaml:148]:
```suggestion
          $ref: '#/components/responses/Generic503'
      security:
        oAuth2AuthorizationCode:
          - "geofencing:subscriptions:read"
```

File: code/API_definitions/geofencing.yaml
Comment (jlurien) [inline code/API_definitions/geofencing.yaml:33]:
```suggestion
      summary: 'Create a geofencing event subscription for a device'
```

**Review by jlurien — COMMENTED** [review]
Add auth2 scopes, some renaming to geofencing

**Comment by jlurien** [comment]
> > Rename api to geofencing
> 
> Hi, I renamed the api two weeks ago before I went to vacation. So I did not recieve any new comment about review, If you think everything is fine you can merge this PR, but first it need two approvers. Anyway I am available if you have any comment. Best, Sadeghi

Hi Sadeghi. I made another review. I think it is almost ready. Once @bigludo7 is back and make a further check we can merge. Thanks!

**Comment by bigludo7** [comment]
@chia2017 @jlurien Thanks for the effort - It looks good for me and thanks to have fixed my previous comments. Once José comments resolved we should be good for merging

**Comment by chia2017** [comment]
I updated the API and applied all last comments that you wrote. If you have no more comment you can merge it and we can close the task. If you have further request you can contact my colleague Maximilian Laue, since I am not working with this project and I switched to the new project.
Best,
Sadeghi

**Comment by maxl2287** [comment]
> I updated the API and applied all last comments that you wrote. If you have no more comment you can merge it and we can close the task. If you have further request you can contact my colleague Maximilian Laue, since I am not working with this project and I switched to the new project. Best, Sadeghi

@bigludo7 & @jlurien:

Could you please review again.
When anything needs to be adapted then I will fix it asap.

Thanks! 🚀

**Comment by bigludo7** [comment]
Thanks @chia2017 for the effort and good luck for your new project
@maxl2287 As of now we cannot approve because following DT suggestion (and CAMARA team approval) the event notification should be aligned with the CloudEvents.
This update is describe here: https://github.com/camaraproject/Commonalities/pull/56  and right now is in team review.

My recommendation is to wait for approval before to adjust your code.
Thanks

**Comment by jlurien** [comment]
Now that https://github.com/camaraproject/Commonalities/pull/56 has been merged, we should update this PR accordingly @maxl2287, @akoshunyadi. Thanks

**Comment by maxl2287** [comment]
I created PR #98 to align on CloudEvents based on these commits here.
In the fact that @chia2017 is not part of the CAMARA project anymore and I am not allowed to add commits to his repository.

**Comment by bigludo7** [comment]
This feature is now followed here: https://github.com/camaraproject/DeviceLocation/pull/98

**Pull request closed without merging**
